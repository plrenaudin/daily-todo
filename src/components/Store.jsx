import React from "react";
import {
  LOAD,
  ADD,
  addTodo,
  REMOVE,
  removeTodo,
  listTodo,
  TOGGLE,
  toggleTodo,
  listWeekTodo
} from "../modules/model";
import { isoFormatDate, currentWeekISO } from "../modules/date";

const ConfigStore = React.createContext();
const TodoStore = React.createContext();
const WeekStore = React.createContext();

const todoReducer = (state, action) => {
  switch (action.type) {
    case LOAD:
      return action.data;
    case ADD:
      return addTodo(state, action.data);
    case REMOVE:
      return removeTodo(state, action.data);
    default:
      throw new Error();
  }
};

const weekReducer = (state, action) => {
  switch (action.type) {
    case LOAD:
      return action.data;
    case TOGGLE:
      return toggleTodo(state, action.todo, action.day);
    default:
      throw new Error();
  }
};

const useDayTodos = () => {
  const [week, dispatch] = React.useReducer(weekReducer, []);
  return { week, dispatch };
};

const useTodos = () => {
  const [todos, dispatch] = React.useReducer(todoReducer, []);
  return { todos, dispatch };
};

const Provider = ({ children }) => {
  //Config
  const [currenDate, setCurrentDate] = React.useState(isoFormatDate());
  const configStore = { currenDate, setCurrentDate };

  //Week
  const weekStore = useDayTodos();

  //Todos
  const todoStore = useTodos();

  React.useEffect(() => {
    const week = currentWeekISO();
    Promise.all([listTodo(), listWeekTodo(week)]).then(([todos, weekTodos]) => {
      todoStore.dispatch({ type: LOAD, data: todos });
      weekStore.dispatch({ type: LOAD, data: weekTodos });
    });
  }, []);

  return (
    <ConfigStore.Provider value={configStore}>
      <TodoStore.Provider value={todoStore}>
        <WeekStore.Provider value={weekStore}>{children}</WeekStore.Provider>
      </TodoStore.Provider>
    </ConfigStore.Provider>
  );
};

export { ConfigStore, WeekStore, TodoStore, Provider };
