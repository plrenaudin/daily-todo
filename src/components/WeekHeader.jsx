import React from "react";
import { currentWeek } from "../modules/date";
import styled from "styled-components";
import { DayHeader } from "./DayHeader";
import { ConfigStore } from "./Store";

const Head = styled.tr`
  padding: 0.3rem;
`;

const WeekHeader = React.memo(() => {
  const config = React.useContext(ConfigStore);
  return (
    <thead>
      <Head>
        {currentWeek(config.currentDate).map(i => (
          <DayHeader key={i.getTime()} day={i} />
        ))}
      </Head>
    </thead>
  );
});

export { WeekHeader };
