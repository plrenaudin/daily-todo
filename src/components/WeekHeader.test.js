import React from "react";
import renderer from "react-test-renderer";
import { WeekHeader } from "./WeekHeader";
import { ConfigStore } from "./Store";

it("renders the WeekHeader correctly", () => {
  const component = (
    <ConfigStore.Provider value={{ currentDate: "2019-02-24" }}>
      <WeekHeader />
    </ConfigStore.Provider>
  );
  expect(renderer.create(component).toJSON()).toMatchSnapshot();
});
