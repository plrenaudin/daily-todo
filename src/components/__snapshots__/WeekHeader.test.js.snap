// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the WeekHeader correctly 1`] = `
.c1 {
  padding: 1rem;
}

.c2 {
  font-size: 1.3rem;
  font-weight: bold;
  display: block;
}

.c3 {
  color: grey;
}

.c0 {
  padding: 0.3rem;
}

<thead>
  <tr
    className="c0"
  >
    <th
      className="c1"
    >
      <em
        className="c2"
      >
        Mo
      </em>
      <small
        className="c3"
      >
        18th
      </small>
    </th>
    <th
      className="c1"
    >
      <em
        className="c2"
      >
        Tu
      </em>
      <small
        className="c3"
      >
        19th
      </small>
    </th>
    <th
      className="c1"
    >
      <em
        className="c2"
      >
        We
      </em>
      <small
        className="c3"
      >
        20th
      </small>
    </th>
    <th
      className="c1"
    >
      <em
        className="c2"
      >
        Th
      </em>
      <small
        className="c3"
      >
        21st
      </small>
    </th>
    <th
      className="c1"
    >
      <em
        className="c2"
      >
        Fr
      </em>
      <small
        className="c3"
      >
        22nd
      </small>
    </th>
    <th
      className="c1"
    >
      <em
        className="c2"
      >
        Sa
      </em>
      <small
        className="c3"
      >
        23rd
      </small>
    </th>
    <th
      className="c1"
    >
      <em
        className="c2"
      >
        Su
      </em>
      <small
        className="c3"
      >
        24th
      </small>
    </th>
  </tr>
</thead>
`;
